c2rust-analyze/src/rewrite/ty.rs

```diff
@@ -0,0 +1,370 @@
+            // TODO: This should generate `OffsetPtr<T>` rather than `&[T]`, but `OffsetPtr` is NYI
+            Quantity::OffsetPtr => tcx.mk_slice(ty),
+        };
+
+        ty = match own {
+            Ownership::Raw => tcx.mk_imm_ptr(ty),
+            Ownership::RawMut => tcx.mk_mut_ptr(ty),
+            Ownership::Imm => tcx.mk_imm_ref(tcx.mk_region(ReErased), ty),
+            Ownership::Cell => tcx.mk_imm_ref(tcx.mk_region(ReErased), ty),
+            Ownership::Mut => tcx.mk_mut_ref(tcx.mk_region(ReErased), ty),
+            Ownership::Rc => todo!(),
+            Ownership::Box => todo!(),
+        };
+
+        ty
+    })
+}
+
+struct HirTyVisitor<'a, 'tcx> {
+    acx: &'a AnalysisCtxt<'a, 'tcx>,
+    rw_lcx: LabeledTyCtxt<'tcx, RewriteLabel>,
+    //mir: &'a Body<'tcx>,
+    //span_index: SpanIndex<Location>,
+    hir_rewrites: Vec<(Span, Rewrite)>,
+}
+
+impl<'a, 'tcx> HirTyVisitor<'a, 'tcx> {
+    fn handle_ty(&mut self, rw_lty: RwLTy<'tcx>, hir_ty: &hir::Ty<'tcx>) {
+        if rw_lty.label.ty_desc.is_none() && !rw_lty.label.descendant_has_rewrite {
+            // No rewrites here or in any descendant of this HIR node.
+            return;
+        }
+
+        let hir_args = match deconstruct_hir_ty(rw_lty.ty, hir_ty) {
+            Some(x) => x,
+            None => {
+                // `hir_ty` doesn't have the expected structure (for example, we expected a type
+                // like `*mut T`, but it's actually an alias `MyPtr`), so we can't rewrite inside
+                // it.  Instead, we discard it completely and pretty-print `rw_lty` (with rewrites
+                // applied).
+                let ty = mk_rewritten_ty(self.rw_lcx, rw_lty);
+                let printer = FmtPrinter::new(*self.rw_lcx, Namespace::TypeNS);
+                let s = ty.print(printer).unwrap().into_buffer();
+                self.hir_rewrites.push((hir_ty.span, Rewrite::PrintTy(s)));
+                return;
+            }
+        };
+
+        if let Some((own, qty)) = rw_lty.label.ty_desc {
+            assert_eq!(hir_args.len(), 1);
+            let mut rw = Rewrite::Sub(0, hir_args[0].span);
+
+            if own == Ownership::Cell {
+                rw = Rewrite::TyCtor("core::cell::Cell".into(), vec![rw]);
+            }
+
+            rw = match qty {
+                Quantity::Single => rw,
+                Quantity::Slice => Rewrite::TySlice(Box::new(rw)),
+                // TODO: This should generate `OffsetPtr<T>` rather than `&[T]`, but `OffsetPtr` is
+                // NYI
+                Quantity::OffsetPtr => Rewrite::TySlice(Box::new(rw)),
+            };
+
+            rw = match own {
+                Ownership::Raw => Rewrite::TyPtr(Box::new(rw), Mutability::Not),
+                Ownership::RawMut => Rewrite::TyPtr(Box::new(rw), Mutability::Mut),
+                Ownership::Imm => Rewrite::TyRef(Box::new(rw), Mutability::Not),
+                Ownership::Cell => Rewrite::TyRef(Box::new(rw), Mutability::Not),
```

Comment:
am I interpreting this as "if something with cell ownership is immutable, rewrite it as an immutable reference" correct?

Reply:
If a pointer `*mut T`/`*const T` has `Ownership::Cell`, we rewrite it into `&Cell<T>`.

>Where would the Cell and mut case go, then?

At this point, we've already decided which pointer type we're rewriting into (as indicated by `own` and `qty`), and we just need to do the rewrite.  The handling of various combinations of `WRITE`/`UNIQUE`/`CELL` permissions and flags is done in `type_desc::perms_to_desc`, which is what computes the `Ownership` and `Quantity` for each pointer type.